_posts/2019-08-23-til-ecto-select-merge.md

```diff
@@ -0,0 +1,237 @@
+---
+author: Kate Travers
+author_link: https://github.com/ktravers
+categories: post
+tags: ['ecto']
+date: 2019-08-23
+layout: post
+title: TIL How to Select Merge with Ecto.Query
+excerpt: >
+  Before you reach for adding another association to your schema, consider using `Ecto.Query#select_merge/3` with a virtual field instead.
+---
+
+Working with Elixir and Ecto, I've run into scenarios where I needed to retrieve data from a table plus maybe a field or two from an unassociated table. In the past, whenever this happened, I'd usually spin up something I wasn't totally satisfied with -- maybe updating the schema(s), breaking it up into multiple queries, or building a multi-select statement if I was feeling fancy.
+
+Happily, today I learned there's a better way. You can accomplish that same end result in a single query expression with `Ecto.Query#select_merge/3`.
+
+Let's run through an example to see it in action.
+
+## Setup
+
+Say you work at a school with an admissions department, and you've been tasked with displaying an event log showing all the events related to a given admission, organized into three columns: 1) date, 2) action taken, and 3) who the action was taken by.
+
+| Date     | Action           | Taken By         |
+|----------|------------------|------------------|
+| 8/1/2019 | Student Admitted | Albus Dumbledore |
+
+
+To start with, we have an `AdmissionEvent` schema that looks like this:
+
+```elixir
+defmodule Registrar.Tracking.AdmissionEvent do
+  use Ecto.Schema
+
+  schema "admission_events" do
+    field(:action, :string)
+    field(:admission_id, :integer)
+    field(:admitter_uuid, :uuid)
```

Comment:
I think you use `uuid` in the migration file and `Ecto.UUID` when defining the schema
so the line should be `field(:admitter_uuid, Ecto.UUID)`

Reply:
i would have thought that too, based on the documentation about [Ecto "custom types"](https://hexdocs.pm/ecto/Ecto.Schema.html#module-custom-types). but my team has an app running in production that defines uuid fields same as above, and we haven't run into any issues. i'm wondering if maybe it has something to do with `Ecto.UUID.type()` being `:uuid`.

that said, if you think folks will find it confusing, i'm happy to change it to `Ecto.UUID`

Reply:
The code didn't compile for me when the field type is `:uuid`
The error I got was: 
```
** (ArgumentError) invalid or unknown type :uuid for field :admitter_uuid
```